modules/checkout/tests/src/Functional/CheckoutOrderTest.php

```diff
@@ -0,0 +1,157 @@
+
+    $variation = ProductVariation::create([
+      'type' => 'default',
+      'sku' => strtolower($this->randomMachineName()),
+      'price' => [
+        'amount' => 9.99,
+        'currency_code' => 'USD',
+      ],
+    ]);
+    $variation->save();
+
+    /** @var \Drupal\commerce_product\Entity\ProductInterface $product */
+    $this->product = Product::create([
+      'type' => 'default',
+      'title' => 'My product',
+      'variations' => [$variation],
+      'stores' => [$store],
+    ]);
+    $this->product->save();
+
+    user_role_grant_permissions(RoleInterface::ANONYMOUS_ID, [
+      'access content',
+      'access checkout',
+    ]);
+  }
+
+  /**
+   * Tests than an order can go through checkout steps.
+   */
+  public function testGuestOrderCheckout() {
+    $this->drupalGet($this->product->toUrl()->toString());
+    $this->submitForm([], 'Add to cart');
+    $cart_link = $this->getSession()->getPage()->findLink('your cart');
+    $cart_link->click();
+    $this->submitForm([], 'Checkout');
+    $this->submitForm([], 'Continue as Guest');
+    $this->submitForm([
+      'contact_information[email]' => 'guest@example.com',
+      'contact_information[email_confirm]' => 'guest@example.com',
+      'billing_information[address][0][recipient]' => $this->randomString(),
+      'billing_information[address][0][organization]' => $this->randomString(),
+      'billing_information[address][0][address_line1]' => $this->randomString(),
+      'billing_information[address][0][locality]' => $this->randomString(),
+    ], 'Continue to review');
+    $this->assertSession()->pageTextContains('Contact information');
+    $this->assertSession()->pageTextContains('Billing information');
+    $this->submitForm([], 'Pay and complete purchase');
+    $this->assertSession()->pageTextContains('Your order number is 1. You can view your order on your account page when logged in.');
+
+    $this->saveHtmlOutput();
+  }
+
+  /**
+   * Creates and sets the default store.
+   *
+   * @return \Drupal\commerce_store\Entity\StoreInterface
+   *    The store.
+   */
+  protected function createDefaultStore() {
+    $currency_importer = \Drupal::service('commerce_price.currency_importer');
+    /** @var \Drupal\commerce_store\StoreStorage $store_storage */
+    $store_storage = \Drupal::service('entity_type.manager')->getStorage('commerce_store');
+
+    $currency_importer->import('USD');
+    $store_values = [
+      'type' => 'default',
+      'uid' => 1,
+      'name' => 'Demo store',
+      'mail' => 'admin@example.com',
+      'address' => [
+        'country_code' => 'US',
+      ],
+      'default_currency' => 'USD',
+      'billing_countries' => [
+        'US',
+      ],
+    ];
+    $store = $store_storage->create($store_values);
+    $store->save();
+    $store_storage->markAsDefault($store);
+
+    return $store;
+  }
+
+  /**
+   * Debugger method to save additional HTML output.
+   */
+  protected function saveHtmlOutput() {
```

Comment:
When using `--printer="\Drupal\Tests\Listeners\HtmlOutputPrinter"` option it writes verbose output. However not on fail. Ending the test with this helped debug errors. We can remove, however, it's been very helpful in writing BrowerTests based tests. It might belong in a CommerceBrowserTestBase if we make one.